src/java.base/share/classes/java/lang/Integer.java

```diff
@@ -547,22 +548,20 @@ public static int parseInt(String s, int radix)
          */
 
         if (s == null) {
-            throw new NumberFormatException("Cannot parse null string");
+            throw NumberFormatException.forNull();
         }
 
         if (radix < Character.MIN_RADIX) {
-            throw new NumberFormatException(String.format(
-                "radix %s less than Character.MIN_RADIX", radix));
+            throw NumberFormatException.forMinRadix(radix);
         }
 
         if (radix > Character.MAX_RADIX) {
-            throw new NumberFormatException(String.format(
-                "radix %s greater than Character.MAX_RADIX", radix));
+            throw NumberFormatException.forMaxRadix(radix);
         }
```

Comment:
Code style should be consistent

Reply:
He means to pull the `radix< Character,MIN_RADIX` and `radix > Character.MAX_RADIX` shared code in Integer and Long.parseInt (with radix version) to a helper method.

Reply:
More explicitly I was thinking something like below. 

I do wonder if some of the benchmark tests should cover the exceptional cases. I have seen many systems where attempting to try and parse potentially invalid inputs can hit expensive NumberFormatException fallbacks (though `-XX:+OmitStackTraceInFastThrow` may eventually kick in). [Guava offers `Integer Ints.tryParse(String)`](https://guava.dev/releases/32.1.3-jre/api/docs/com/google/common/primitives/Ints.html#tryParse(java.lang.String)) and similar to avoid the exception overhead at the cost of boxing for happy path of valid `int` and `null` for invalid.

```
        checkCanParseLength(s);
        checkCanParseRadix(radix);
        int len = s.length();


    private static void checkCanParseLength(String s) {
        if (s == null) {
            throw NumberFormatException.forNull();
        }

        if (s.isEmpty()) { // would be interesting to see if this is equivalent to `s.length() == 0` for happy paths
            throw NumberFormatException.forEmpty();
        }
    }

    private static void checkCanParseRadix(int radix) {
        if (radix < Character.MIN_RADIX) {
            throw NumberFormatException.forMinRadix(radix);
        }

        if (radix > Character.MAX_RADIX) {
            throw NumberFormatException.forMaxRadix(radix);
        }
    }
```

Reply:
In addition to FreqInlineSize 325, there are other reasons that affect inlining. If inlining fails, there will be one more method call, making it slower.

Reply:
This PR does not optimize the scenario where inputting illegal values leads to exceptions.

Reply:
This patch should have little impact on exception-path performance. Remember String concatenation is done with StringBuilder in java.base, so changing formatter to that might have some performance difference. However, the main overhead of exception was from creation and initialization of stack trace.; the new code was just copies of the existing generic-radix parseInt, except it has a more efficient way to obtain digits via a table.

Reply:
Let's not forget that `Character.digit()` has a pretty decent fast path for Latin1, including a 256 bytes lookup-table.

Reply:
If we reuse CharacterDataLatin1#DIGITS, the performance will be slower, the performance numbers are as follows

```java
class CharacterDataLatin1 {
    static int digit(int ch) {
        int value = DIGITS[ch & 0xFF];
        return (value >= 0 && value < 10) ? value : -1;
    }
}
```

```diff
Benchmark          (size)  Mode  Cnt  Score   Error  Units (baseline)
Integers.parseInt     500  avgt   15  2.614 ? 0.003  us/op

Benchmark          (size)  Mode  Cnt  Score   Error  Units (use CharacterDataLatin1#digit)
Integers.parseInt     500  avgt   15  2.399 ? 0.008  us/op

Benchmark          (size)  Mode  Cnt  Score   Error  Units (c3e878a)
Integers.parseInt     500  avgt   15  2.251 ? 0.010  us/op
```

Reply:
Maybe there's something wrong with the `value < radix` check that JIT cannot efficiently optimize it (so `digit <= 9` has a similar performance to `value < 10` in loops). We might have to check the hotspot assembly to find out the reason.